Aa/pygame project (#2)
Ada-Developers-Academy/core-projects: pygame-blob/intro.md

# Pygame Blob

## Skills Assessed
- Solving problems with… 
    - Creating Classes
    - Create instances 
    - Using inheritance
    - Overriding methods 
    - Importing modules 
    - Using dunder methods

## Goals
Python is such a versatile language. We can create APIs, do data analysis, build web apps, and create games! In this project, we will use OOP to build an animation in pygame. 

**What is pygame?**

Pygame is a library that allows developers to create games using Python.
To provide a better visual picture of OOP, we’ll be creating an animation of blobs! 

(Insert gif of finished project)

## Project

> Pyblobs will be a **2-day project**. The first day will consist of a tutorial which can be done independently. The second day will consist of working in pairs to extend the project made from the tutorial.

Reviewer: ```suggestion
Pyblobs is a **2-day project**. The first day you will make an animation in pygame by completing a tutorial independently. The second day we will review the content of the tutorial together in round tables and then you will work in pairs to extend and enhance your animations.
```

## Tutorial

> The tutorial will be done independently on day 1 of the project. We'll follow this tutorial on [pythonprogramming.net](https://pythonprogramming.net/object-oriented-programming-introduction-intermediate-python-tutorial/). 

Reviewer: ```suggestion
We will follow this tutorial on [pythonprogramming.net](https://pythonprogramming.net/object-oriented-programming-introduction-intermediate-python-tutorial/). 
```

## Initial Set Up

> To create our game we will use **repl.it**, a browser based IDE. Repl allows developers to create and run mini applications in the browser. We highly suggest [creating an account](https://repl.it/signup) with Repl using your github credentials. This will allow you to create repositories and commits to Github using Repl's interface. 

Reviewer: Do we actually want to use the github commits feature? Is this what's in beta?

For a more collaborative experience, we also suggest using repl's multiplayer feature to have you and your partner edit on the same repl together. To use multiplayer mode, click "Share" on the top right, a pop-up window should appear with the invite tab open. Type in your partners username and click 'invite'. 

## Getting Started with Pygame 

> 1) Download the Pygame library in the packages section. 
> 2) Once you see a `poetry.lock` and `pyproject.toml` file appear in the Packager files section on repl.it you are ready to start the tutorial!

Reviewer: ```suggestion
1) Create a New Repl
2) Download the Pygame library in the packages section (accessed in the lefthand menu). 
3) Once you see a `poetry.lock` and `pyproject.toml` file appear in the Packager files section on repl.it you are ready to start the tutorial!
```

## Tutorial
We will complete 5 lessons from this [tutorial](https://pythonprogramming.net/object-oriented-programming-introduction-intermediate-python-tutorial/) as outlined below.

### Create one single moving blob
- [Introduction to OOP](https://pythonprogramming.net/object-oriented-programming-introduction-intermediate-python-tutorial/)
- [Creating Environment for Object with Pygame](https://pythonprogramming.net/creating-pygame-environment-intermediate-python-tutorial/)

### Create many moving blobs
- [Many blobs - OOP](https://pythonprogramming.net/many-blob-objects-intermediate-python-tutorial/)
- [Blob Class and Modularity OOP](https://pythonprogramming.net/class-object-modularity-intermediate-python-tutorial/)

### Reflection Question 1: `check_bounds`
At the end of the lesson, the author has this to say about the `check_bounds` method: "Now, the programmer can decide to use it or not. You could also give some sort of argument in the move method, where, if `True`, then boundaries would be enforced."

> Consider how your would implement this optional enforcement of boundary checking, and record your thoughts here.

Reviewer: <!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: short-answer
* id: be93d3ad-7206-42e2-b00a-d1fcd4e3a1ab
* title: Reflection Question 1: `check_bounds`
<!-- * points: [1] (optional, the number of points for scoring as a checkpoint) -->
<!-- * topics: [python, pandas] (optional the topics for analyzing points) -->

##### !question

At the end of the lesson, the author has this to say about the `check_bounds` method: "Now, the programmer can decide to use it or not. You could also give some sort of argument in the move method, where, if `True`, then boundaries would be enforced."

Consider how your would implement this optional enforcement of boundary checking, and record your thoughts here.

##### !end-question

##### !placeholder

Input your answer here

##### !end-placeholder

##### !answer

/*+/

##### !end-answer

<!-- other optional sections -->
<!-- !hint - !end-hint (markdown, hidden, students click to view) -->
<!-- !rubric - !end-rubric (markdown, instructors can see while scoring a checkpoint) -->
<!-- !explanation - !end-explanation (markdown, students can see after answering correctly) -->

### !end-challenge

<!-- ======================= END CHALLENGE ======================= -->

Reviewer: Not sure if the Learn markdown translated correctly here.

- [Inheritance - OOP](https://pythonprogramming.net/inheritance-object-oriented-programming-intermediate-python-tutorial/)

> - ### Reflection Question 2: `move` vs `move_fast`
> 
> The author changes the name of the `move` method to `move_fast` in the child class `BlueBlob`. Is this name change necessary? What would happen if we left it as `move`?

Reviewer: ```suggestion
<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: short-answer
* id: be93d3ad-7206-42e2-b00a-d1fcd4e3a1ab
* title: Reflection Question 2: `move` vs `move_fast`
<!-- * points: [1] (optional, the number of points for scoring as a checkpoint) -->
<!-- * topics: [python, pandas] (optional the topics for analyzing points) -->

##### !question

The author changes the name of the `move` method to `move_fast` in the child class `BlueBlob`. Is this name change necessary? What would happen if we left it as `move`?

##### !end-question

##### !placeholder

Input your answer here

##### !end-placeholder

##### !answer

/*+/

##### !end-answer

<!-- other optional sections -->
<!-- !hint - !end-hint (markdown, hidden, students click to view) -->
<!-- !rubric - !end-rubric (markdown, instructors can see while scoring a checkpoint) -->
<!-- !explanation - !end-explanation (markdown, students can see after answering correctly) -->

### !end-challenge

<!-- ======================= END CHALLENGE ======================= -->
```

### Optional Reading (No more blobs, but some fun python OOP concepts)
- [Decorators in Python](https://pythonprogramming.net/decorators-intermediate-python-tutorial/) 
- [Special Methods, OOP, and Iteration Python](https://pythonprogramming.net/special-methods-iteration-intermediate-python-tutorial/)

> ## Pair Project

Reviewer: ```suggestion
## Pair Project - Day 2
```
